File: core/src/main/java/heast/core/network/pipeline/ClientConnection.java
package heast.core.network.pipeline;

import heast.core.logging.IO;
import heast.core.network.NetworkSide;
import heast.core.network.NetworkState;
import heast.core.network.Packet;
import heast.core.network.PacketListener;
import io.netty.bootstrap.Bootstrap;
import io.netty.channel.*;
import io.netty.channel.nio.NioEventLoopGroup;
import io.netty.channel.socket.SocketChannel;
import io.netty.channel.socket.nio.NioSocketChannel;
import io.netty.util.AttributeKey;

import javax.crypto.SecretKey;

public final class ClientConnection extends SimpleChannelInboundHandler<Packet<?>> {

    public static final AttributeKey<NetworkState> PROTOCOL_KEY = AttributeKey.valueOf("protocol");

    private final NetworkSide side;
    private String verificationCode;
    private Runnable onVerifySuccess;
    private Channel channel;
    private PacketListener listener;

    public ClientConnection(NetworkSide side) {
        this.side = side;
    }

    public NetworkSide getSide() {
        return side;
    }

    public void setVerificationCode(String verificationCode) {
        this.verificationCode = verificationCode;
    }

    public String getVerificationCode() {
        return verificationCode;
    }

    public void setOnVerifySuccess(Runnable onVerifySuccess) {
        this.onVerifySuccess = onVerifySuccess;
    }

    public void onVerifySuccess() {
        onVerifySuccess.run();
    }

    public Channel getChannel() {
        return channel;
    }

    public void setState(NetworkState state) {
        channel.attr(PROTOCOL_KEY).set(state);
    }

    public NetworkState getState() {
        return channel.attr(PROTOCOL_KEY).get();
    }

    public void setListener(PacketListener listener) {
        this.listener = listener;
    }

    public PacketListener getListener() {
        return listener;
    }

    @Override
    public void channelActive(ChannelHandlerContext ctx) {
        channel = ctx.channel();
        setState(NetworkState.LOGIN);
    }

    @Override
    public void channelInactive(ChannelHandlerContext ctx) {
        if (channel.isOpen()) {
            channel.close().awaitUninterruptibly();
        }
    }

    @Override
    protected void messageReceived(ChannelHandlerContext ctx, Packet<?> msg) {
        handle(msg, listener);
    }

    @SuppressWarnings("unchecked")
    private static <T extends PacketListener> void handle(Packet<T> packet, PacketListener listener) {
        try {
            packet.apply((T) listener);
        } catch (Exception e) {
            System.err.println("ClientConnection: Error whilst handling packet, make sure to use the correct packet listener class");
            e.printStackTrace();
        }
    }

    public void send(Packet<?> packet, Runnable onSent) {
        sendInternal(packet, onSent);
    }

    public void send(Packet<?> packet) {
        sendInternal(packet, () -> {});
    }

    private void sendInternal(Packet<?> packet, Runnable onSent) {
        if (channel != null) {
            channel.writeAndFlush(packet)
                .addListener( future -> {
                    if (!future.isSuccess()) {
                        System.err.println("ClientConnection: Failed to send packet '" + packet.getClass().getSimpleName() + "'");
                        future.cause().printStackTrace();
                    } else {
                        onSent.run();
                    }
                });
        } else {
            throw new NullPointerException(
                "ClientConnection: Failed to send packet, the channel is not available"
            );
        }
    }

    public void enableEncryption(SecretKey key) {
        channel.pipeline().addBefore("encoder", "encryptor", new PacketEncryptor(key));
        channel.pipeline().addBefore("decoder", "decryptor", new PacketDecryptor(key));
    }

    public static ClientConnection connect(String host, int port) {
        var connection = new ClientConnection(NetworkSide.CLIENT);
        var workerGroup = new NioEventLoopGroup();
        new Bootstrap()
            .group(workerGroup)
            .handler(new ChannelInitializer<SocketChannel>() {
                public void initChannel(SocketChannel ch) {
                    ch.config().setOption(ChannelOption.TCP_NODELAY, true);
                    ch.pipeline()
                        // Here will be the packet decryptor
                        .addLast("decoder", new PacketDecoder(NetworkSide.SERVER))
                        // Here will be the packet encryptor
                        .addLast("encoder", new PacketEncoder(NetworkSide.CLIENT))
                        .addLast("handler", connection);
                }
            })
            .option(ChannelOption.TCP_NODELAY, true)
            .option(ChannelOption.SO_KEEPALIVE, true)
            .channel(NioSocketChannel.class)
            .connect(host, port).addListener(future -> {
                if (future.isSuccess()) {
                    IO.info.println("Connected to server");
                } else{
                    IO.error.println("Failed to connect to server: " + future.cause());
                }})
            .syncUninterruptibly();
        return connection;
    }
}

File: Core/network/pipeline/ClientConnection.cs
using Core.exceptions;
using DotNetty.Common.Utilities;
using DotNetty.Transport.Bootstrapping;
using DotNetty.Transport.Channels;
using DotNetty.Transport.Channels.Sockets;

namespace Core.network.pipeline; 

public class ClientConnection : SimpleChannelInboundHandler<IPacket<IPacketListener>> {

    public static AttributeKey<NetworkState> ProtocolKey => AttributeKey<NetworkState>.ValueOf("protocol");

    public NetworkSide Side { get; }
    public IChannel? Channel { get; private set; }
    public IPacketListener? Listener { get; private set; }

    public ClientConnection(NetworkSide side) {
        this.Side = side;
    }
    
    public static Task GetConnection(string host, int port) {
        var connection = new ClientConnection(NetworkSide.Client);
        var workerGroup = new MultithreadEventLoopGroup();

        return new Bootstrap()
            .Group(workerGroup)
            .Handler(new ClientConnectionInitializer(connection))
            .Option(ChannelOption.TcpNodelay, true)
            .Option(ChannelOption.SoKeepalive, true)
            .Channel<TcpSocketChannel>()
            .ConnectAsync(host, port);
    }
    
    public void SetState(NetworkState state) {
        Channel?.GetAttribute(ProtocolKey).Set(state);
    }
    
    public NetworkState? GetState() { 
        return Channel?.GetAttribute(ProtocolKey).Get();
    }
    
    public void SetListener(IPacketListener listener) {
        this.Listener = listener;
    }
    
    public IPacketListener? GetListener() {
        return this.Listener;
    }

    public override void ChannelActive(IChannelHandlerContext context) {
        Channel = context.Channel;
        SetState(NetworkState.Login);
    }

    public override async void ChannelInactive(IChannelHandlerContext context) {
        if (Channel is { Open: true }) {
            await Channel.CloseAsync();
        }
    }

    protected override void ChannelRead0(IChannelHandlerContext ctx, IPacket<IPacketListener> msg) {
        if (Listener == null)
            throw new IllegalStateException("Listener was null whilst a message was received");

        msg.Apply(Listener);
    }
    
    public Task Send(IPacket<IPacketListener> packet) {
        return Channel is { Open: true } ?
            Channel.WriteAndFlushAsync(packet) :
            throw new IllegalStateException("Channel was null whilst trying to send a packet");
    }
}

public class ClientConnectionInitializer : ChannelInitializer<ISocketChannel> {
    private ClientConnection Connection { get; }
    
    public ClientConnectionInitializer(ClientConnection connection) {
        Connection = connection;
    }
    
    protected override void InitChannel(ISocketChannel channel) {
        channel.Configuration.SetOption(ChannelOption.TcpNodelay, true);
        channel.Pipeline
            // Here will be the packet decryptor
            .AddLast("decoder", new PacketDecoder(NetworkSide.Server))
            // Here will be the packet encryptor
            .AddLast("encoder", new PacketEncoder(NetworkSide.Client))
            .AddLast("handler", Connection);
    }
}